Respect peer dependencies

korelstar opened the pull request:
The purpose of this PR is to fix #376. 

@raineorshine Please have a look if this goes in the correct direction. Open questions are, e.g.:
- should the check be optional, what is default?
- which warning should be raised if detection does not work? (the tests are currently failing due to this warning)

I've exported the new `getPeerDependencies` for the purpose that other tools can use it too, e.g. https://github.com/th0r/npm-upgrade

raineorshine commented on lib/index.js:
```diff
@@ -224,6 +228,32 @@ async function analyzeProjectDependencies(options, pkgData, pkgFile) {
+      if (acc[pkgName] === undefined) {
+        acc[pkgName] = []
+      }
+      // eslint-disable-next-line fp/no-mutating-methods
```
Note: PR's that disable the linter will not be accepted in order to maintain the project's code style and quality requirements. You'll have to find a way to do this in a more functional style with no mutations.

korelstar replied:
To be honest, I just copied that from your own code: https://github.com/raineorshine/npm-check-updates/search?q=no-mutating-methods

But not problem, I've fixed that now :-)

raineorshine replied:
Fair enough! There is some legacy code from an older version that had different linter settings.

raineorshine reviewed (changes requested):
> The purpose of this PR is to fix #376.
> 
> @raineorshine Please have a look if this goes in the correct direction.

Thanks for your contribution!

> * should the check be optional, what is default?

It should be opt-in via the `--peer` option. Default `ncu` behavior is intentionally kept very minimal so it is easy to reason about and there are no surprises.

> * which warning should be raised if detection does not work? (the tests are currently failing due to this warning)

Can you explain more what you mean by "does not work"? Do you mean an upgrade breaks a peer dependency (expected), or do you mean the detection algorithm doesn't work in some cases (unexpected)? Are there other cases besides when packages have not yet been installed that you are thinking of?

Some other thoughts:

- Should `ncu` ignore breaking peer dependencies completely when this option is provided, or should it suggest the latest version that is compatible? Not sure what would be expected/most useful.
- Unit test required

Thanks!

korelstar commented:
> It should be opt-in via the `--peer` option. Default `ncu` behavior is intentionally kept very minimal so it is easy to reason about and there are no surprises.

The `--peer` option can't be used, since there is already an attribute `peer` in the `options` variable which is `true` by default. It looks like it is filled by the `--dep` command line option. Therefore, I have chosen the option `--checkPeer` for now.

> > * which warning should be raised if detection does not work? (the tests are currently failing due to this warning)
> 
> Can you explain more what you mean by "does not work"? Do you mean an upgrade breaks a peer dependency (expected), or do you mean the detection algorithm doesn't work in some cases (unexpected)? Are there other cases besides when packages have not yet been installed that you are thinking of?

In fact, I think it's only relevant when a package is not installed locally. An upgrade should never break a peer dependency, since the purpose of this PR is to check this (see next comment).

> * Should `ncu` ignore breaking peer dependencies completely when this option is provided, or should it suggest the latest version that is compatible? Not sure what would be expected/most useful.

My goal is that `ncu` suggests the latest version that is compatible to the peer dependencies. I think this is what a user would expect and this is what I've implemented. I don't think a user would expect that peer dependency are ignored completely.

> * Unit test required

I've added a test, now. The test uses a package [`ncu-test-peer`](https://www.npmjs.com/package/ncu-test-peer) (which has a peer dependency to `ncu-test-return-version`) for this purpose. We can transfer the ownership to you, if you want.

raineorshine commented:
> > It should be opt-in via the `--peer` option. Default `ncu` behavior is intentionally kept very minimal so it is easy to reason about and there are no surprises.
> 
> The `--peer` option can't be used, since there is already an attribute `peer` in the `options` variable which is `true` by default. It looks like it is filled by the `--dep` command line option. Therefore, I have chosen the option `--checkPeer` for now.

Perfect! I will go ahead and remove those options since they are legacy and internal only.

> > * Should `ncu` ignore breaking peer dependencies completely when this option is provided, or should it suggest the latest version that is compatible? Not sure what would be expected/most useful.
> 
> My goal is that `ncu` suggests the latest version that is compatible to the peer dependencies. I think this is what a user would expect and this is what I've implemented. I don't think a user would expect that peer dependency are ignored completely.

That makes sense.

> > * Unit test required
> 
> I've added a test, now. The test uses a package [`ncu-test-peer`](https://www.npmjs.com/package/ncu-test-peer) (which has a peer dependency to `ncu-test-return-version`) for this purpose. We can transfer the ownership to you, if you want.

Yes! That would be great.

raineorshine commented on lib/index.js:
```diff
@@ -224,6 +228,31 @@ async function analyzeProjectDependencies(options, pkgData, pkgFile) {
+    try {
+      const pkgData = fs.readFileSync(path)
+      const pkg = jph.parse(pkgData)
+      return vm.getCurrentDependencies(pkg, { dep: 'peer', filter: options.filter, reject: options.reject })
```
We should probably pass the all of the users' options and just override `dep: 'peer'`:

```js
return vm.getCurrentDependencies(pkg, { ...options, dep: 'peer' })
```

korelstar replied:
Yes, that sounds good. I've fixed it.

korelstar commented:
> > > * Unit test required
> > 
> > 
> > I've added a test, now. The test uses a package [`ncu-test-peer`](https://www.npmjs.com/package/ncu-test-peer) (which has a peer dependency to `ncu-test-return-version`) for this purpose. We can transfer the ownership to you, if you want.
> 
> Yes! That would be great.

I've added you as maintainer. We can remove me now or after this PR is merged. Whatever you prefer.

raineorshine commented:
Published in `v11.4.0`
